### PR TITLE
refactor(alert): remove intermediate 'div' element

### LINKED ISSUE #2544: alert: remove superfluous <div> tag
As of today `<ngb-alert>` doesn't set Bootstrap classes on a host but [adds a `<div>` inside](https://github.com/ng-bootstrap/ng-bootstrap/blob/e0c005082f11f811ecc6e723e27ec029eb29770a/src/alert/alert.ts#L18). This is an issue for animations and prevents people from properly styling the alert's host (see https://stackoverflow.com/questions/51450796/apply-classes-to-sub-elements-of-components/51451075 and #2542 as an example). 

We should remove [this `<div>`](https://github.com/ng-bootstrap/ng-bootstrap/blob/e0c005082f11f811ecc6e723e27ec029eb29770a/src/alert/alert.ts#L18) and set classes (and `role="alert"` directly on the host).

### PR DESCRIPTION
**Pull request by maxokorokov**
Fixes #2544

### CODE REVIEW

File: src/alert/alert.ts [hunk 1]
```diff
@@ -38,12 +43,22 @@ export class NgbAlert {
...
     this.dismissible = config.dismissible;
     this.type = config.type;
   }
 
   closeHandler() { this.close.emit(null); }
+
+  ngOnChanges(changes: SimpleChanges) {
+    const {type} = changes;
+    if (type && !type.firstChange) {
+      this._renderer.removeClass(this._element.nativeElement, `alert-${type.previousValue}`);
```
Comment (pkozlowski-opensource) [inline src/alert/alert.ts]: Why do we do this through low-level DOM manipulation? Couldn't we move the existing bindings to the host?
Or is binding overriding classes set by a user on a host?

Comment (maxokorokov) [reply src/alert/alert.ts]: Double-checked, and it overrides user-set classes. So I see three options:

1. Leave as it is done currently with `Renderer2`
2. hardcode all supported classes and remove the demo with `alert-customclass`:

```
  host: {
    'role': 'alert',
    'class': 'alert d-block',
    '[class.alert-dismissible]': 'dismissible',
    '[class.alert-success]': `type === 'success'`,
    '[class.alert-warning]': `type === 'warning'`,
    '[class.alert-info]': `type === 'info'`,
    '[class.alert-danger]': `type === 'danger'`,
    '[class.alert-primary]': `type === 'primary'`,
    '[class.alert-secondary]': `type === 'secondary'`,
    '[class.alert-light]': `type === 'light'`,
    '[class.alert-dark]': `type === 'dark'`
  },
```
3. do nothing and close this :)

File: src/alert/alert.ts [hunk 2]
```diff
@@ -38,12 +43,22 @@ export class NgbAlert {
...
-  constructor(config: NgbAlertConfig) {
+  constructor(config: NgbAlertConfig, private _renderer: Renderer2, private _element: ElementRef) {
     this.dismissible = config.dismissible;
     this.type = config.type;
   }
 
   closeHandler() { this.close.emit(null); }
+
+  ngOnChanges(changes: SimpleChanges) {
+    const {type} = changes;
```
Comment (pkozlowski-opensource) [inline src/alert/alert.ts]: While it works I find the `type` name confusing / not very descriptive  here. A much better name would be `typeChange`. I would prefer to be explicit here.

File: src/alert/alert.ts [hunk 3]
```diff
@@ -38,12 +43,22 @@ export class NgbAlert {
...
+    if (type && !type.firstChange) {
+      this._renderer.removeClass(this._element.nativeElement, `alert-${type.previousValue}`);
+      this._renderer.addClass(this._element.nativeElement, `alert-${type.currentValue}`);
+    }
+  }
+
+  ngOnInit() { this._renderer.addClass(this._element.nativeElement, `alert-${this.type}`); }
```
Comment (pkozlowski-opensource) [inline src/alert/alert.ts]: Couldn't we move this logic to `ngOnChanges` and set the class on `firstChange`? This way the whole logic would be centralised in one place

**Comment by pkozlowski-opensource** [comment]
@maxokorokov I left few nits but generally speaking I think that we should merge it. It is not an ideal solution but this is the best we've got given current Angular...

File: src/alert/alert.ts (hunk 2 above)
Comment (maxokorokov) [reply src/alert/alert.ts]: Will change

**Comment by maxokorokov** [comment]
@pkozlowski-opensource , fixed comments and:
- added a `BREAKING CHANGE` comment about the markup change
- replaced `d-block` with `:host { display: block; }` as it doesn't set `!important` flag

**Pull request merged**